### DISCUSSION TITLE
ui.notification_bar

### DISCUSSION THREAD

**Discussion by azim-b** [h2oai/wave#1431] [Q&A]
**Wave version: h2o_wave-nightly-py3-none-macosx_10_9_x86_64.whl**
-----

1. redirect and notification bar don't work together

### **without** redirect

https://user-images.githubusercontent.com/64787868/169142858-f803496c-ad66-41b5-8900-12f218701295.mov

Code:
```
async def notify(q:Q):
    q.page['meta'] = ui.meta_card(box='', 
    notification_bar=ui.notification_bar(
        text='You have 1 new notification',
        type='success',
        position='top-right',
    ))
```

### **with** redirect

https://user-images.githubusercontent.com/64787868/169142876-27029f7c-4662-4453-ab79-7330b55eb36a.mov

Code:
```
async def notify(q:Q):
    q.page['meta'] = ui.meta_card(box='', 
    # REDIRECT
    redirect='/',
    notification_bar=ui.notification_bar(
        text='You have 1 new notification',
        type='success',
        position='top-right',
    ))
```

---

2. Notification bar close (`X`) button.

The close / `X` button only "temporarily" closes the notification bar, that is, the notification bar "reappears" when the form is submitted again. If this is intentional and the developer is required to explicitly close it (`q.page['meta'].notification_bar = None`), this requires adding and handling another button, which seems redundant. Is there a way to handle the close button click? Tried setting the `events` param but that failed and it's not fully clear what its purpose is.

https://user-images.githubusercontent.com/64787868/169144424-09ca4bdb-f15c-45c7-8d52-e9912f6e4160.mov

**Comment by mturoci** [comment]
> 1

It's odd to have both redirect and notification_bar set together as it's effectively saying: Display notification bar and go to a new URL at the same time. Since the redirect is handled sooner, the page just redirects and the notification bar is left in the previous page. What you most likely want is to redirect and show a notification afterwards:

```py
async def notify(q:Q):
    q.page['meta'] = ui.meta_card(box='')
    # REDIRECT
    q.page['meta'].redirect = '/'
    # Push the redirect changes to the browser.
    await q.page.save()
    # Show notification.
    q.page['meta'].notification_bar=ui.notification_bar(
        text='You have 1 new notification',
        type='success',
        position='top-right',
    ))
```
> 2

`dismissed` event was added in https://github.com/h2oai/wave/issues/1386. Should work the same as dialog/panel. Available in nightly so feel free to try it out.

**Reply by azim-b** [reply]
> What you most likely want is to redirect and show a notification afterwards

^ this is what I would like to achieve, and I tried it too (my own code), but failed. The reproducible code that you sent has a similar patter to that of my code that I tried earlier. Is your repro code producing the desired result?

**Reply by mturoci** [reply]
>Is your repro code producing the desired result?

The code was to provide a general idea. One thing to be aware of is that `serve` function is called again after redirect which means your `notify` handler will most likely not get called as `q.args.notify` will not be present. The minimal working code:

```py
from h2o_wave import main, app, Q, ui


@app('/')
async def serve(q: Q):
    q.page['card'] = ui.form_card(box='1 1 2 2', items=[
        ui.button(name='btn', label='Click'),
    ])

    q.page['meta'] = ui.meta_card(box='')
    if q.args.btn:
        # REDIRECT
        q.page['meta'].redirect = '/'
        # Push the redirect changes to the browser.
        await q.page.save()
        # await q.sleep(2)
    else:
        q.page['meta'].notification_bar = ui.notification_bar(
            text='You have 1 new notification',
            type='success',
            position='top-right',
        )

    await q.page.save()
```
